community/kernel/src/main/java/org/neo4j/kernel/api/schema_new/OrderedPropertyValues.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.schema_new;
+
+import java.util.Arrays;
+
+/**
+ * Holder for n property values, ordered according to a schema descriptor property id order
```

Comment:
The property value order in not enforced yet in this PR, and the places that throw IndexEntryConflictExceptions (`UniqueInMemoryIndex`, `DuplicateCheckingCollector` and `ConstraintEnforcingEntityOperations`) do not yet support composite properties. We broke this exception part out into a separate from need in implementing these parts. I think providing a single point of enforcement might be clumsy, and would rather leave that to callers.